### ISSUE TITLE
Elastic collisions?

### ISSUE THREAD

**Issue by rschwa6308** [rschwa6308/Physics-2.0#14]
Currently all collisions are perfectly inelastic, should functionality for choosing any arbitrary level of elasticity be added?

**Comment by StardustGogeta** [comment]
That sounds pretty interesting, but I am not sure it would seem very realistic to have planets and stars bouncing around like balls on a pool table. 😃 Are you concerned about realism, or is that not much of a priority?

**Comment by rschwa6308** [comment]
That would be hilarious! While it might not be realistic, I think it might be worth it to add.  Perfectly elastic collisions = conservation of energy, so we will likely get interesting behavior from the simulations.  I was thinking of adding another slider labeled "elasticity" that ranges from 0 to 1.

**Comment by StardustGogeta** [comment]
Okay, I see. How exactly would you quantify the elasticity when it is neither perfectly elastic nor perfectly inelastic? Would you take away a percentage of the difference in overall kinetic energy (between elastic and inelastic) while still maintaining conservation of momentum? I can't quite remember how to do something to an arbitrary scale like that. 😕

**Comment by rschwa6308** [comment]
The slider will represent the Coefficient of restitution for all collisions.  

e = sqrt(KE'/KE)

e will be directly proportional to the ratio of the velocities before and after the collision.

**Comment by StardustGogeta** [comment]
That's perfect! I can't believe I have never heard of that before. It would be neat to turn the coefficient up beyond 1 to have planets appear to explode against each other!

**Comment by rschwa6308** [comment]
It took a good bit of googling for me to find out what that quantity is actually called XD. e > 1 sounds very exciting :)

**Comment by rigille** [comment]
I actually modified your program to make elastic collisions some time ago.
The intent was to simulate the rings of Saturn. It didn't end up very realistic --- as you might guess. It was fun nevertheless. :)

**Comment by StardustGogeta** [comment]
@rschwa6308 In my [test branch](https://github.com/StardustGogeta/Physics-2.0/tree/elasticity), I added a start setting for just planets and no stars, and I set the masses to be equal to resemble billiards. It appeared to be correct there, when I tested it. Do you still notice any issues with it?

Below, I have put some examples I used to test the calculations with the COR set to 1. They are all correct in these demonstrations.

```>>> m,m2,v,v2= 1,1,V2(1,0), V2(0,1)
>>> (m*v + m2*v2 + m2*COR*(v2-v)) / (m+m2)
<Vector2(0, 1)>
>>> (m*v + m2*v2 + m*COR*(v-v2)) / (m+m2)
<Vector2(1, 0)>
>>> m,m2,v,v2= 1,1,V2(1,0), V2(0,0)
>>> (m*v + m2*v2 + m2*COR*(v2-v)) / (m+m2)
<Vector2(0, 0)>
>>> (m*v + m2*v2 + m*COR*(v-v2)) / (m+m2)
<Vector2(1, 0)>
>>> m,m2,v,v2= 1,2,V2(1,0), V2(0,0)
>>> (m*v + m2*v2 + m2*COR*(v2-v)) / (m+m2)
<Vector2(-0.333333, 0)>
>>> (m*v + m2*v2 + m*COR*(v-v2)) / (m+m2)
<Vector2(0.666667, 0)>

**Comment by rschwa6308** [comment]
It all looks correct to me! I guess the math does work! The issue I'm seeing is that when bodies collide, they simple reverse their velocities, instead of bouncing off at arbitrary angles.  I think this is due to the fact that bodies are just points, and a 'collision' occurs when the distance between two points is less than the sum of the radii.  I think we need to add some calculations to determine the resulting angles of the bodies' velocities.

**Comment by StardustGogeta** [comment]
Yes, that's it, you're right! Every collision is calculated as the centers colliding rather than the edges of the circles themselves. That must be why it is so much more noticeable with the stars and their extremely large radii.

**Comment by rschwa6308** [comment]
I updated my variable-elasticty branch to fix this issue, but it doesn't incorporate CoR properly yet

**Comment by StardustGogeta** [comment]
I am trying to take a look at it now, but I am having issues with merging everything together. 😒

**Comment by StardustGogeta** [comment]
Okay, I got all the merging fixed and proposed pull request #15 to fix any differences we still have between our forks.
